### PR TITLE
chore(Analytics): migrate key-value store from NSUserdefaults to keychain

### PR DESCRIPTION
**Pull request by phantumcode**
*Issue #, if available:*
*Description of changes:*
* Update AWSPinpoint to migrate key-value storage to Keychain storage for the following key-values: `AWSPinpointEndpointAttributesKey`, `AWSPinpointEndpointMetricsKey`, `AWSPinpointEndpointProfileKey`, `com.amazonaws.AWSDeviceTokenKey`

* Updated unit tests and integration tests 
* Updated AWSPinpointTest to use a Test Host Application in order to read/write to keychain in integration tests

*Check points:*

- [x] Added new tests to cover change, if needed
- [x] All unit tests pass
- [x] All integration tests pass
- [x] Updated CHANGELOG.md
- [x] Documentation update for the change if required
- [x] PR title conforms to conventional commit style

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by ruisebas — CHANGES_REQUESTED** [review]
The `AWSDeviceTokenKey` value is first stored in UserDefaults on **AWSPinpointNotificationManager's** [`interceptDidRegisterForRemoteNotificationsWithDeviceToken`](https://github.com/aws-amplify/aws-sdk-ios/blob/1717c48d194406f193bf0737b9b3106655f1ce39/AWSPinpoint/AWSPinpointNotificationManager.m#L114-L125), so we should address it as well.

I think you might want to update it so that it directly store in the keychain instead, and invert the condition in `AWSPinpointEndpointProfile`'s `init` (so that it first checks the Keychain, and then UserDefaults). But up to you

File: AWSPinpoint/AWSPinpointEndpointProfile.m [hunk 1]
```diff
@@ -64,22 +65,35 @@ - (instancetype) initWithApplicationId:(NSString*) applicationId
...
                 applicationLevelOptOut:(BOOL) applicationLevelOptOut
                                  debug:(BOOL) debug
-                          userDefaults:(NSUserDefaults*) userDefaults {
+                          userDefaults:(NSUserDefaults*) userDefaults
+                              keychain: (AWSUICKeyChainStore*) keychain {
     if (self = [super init]) {
-        NSData *tokenData = [userDefaults objectForKey:AWSDeviceTokenKey];
-        NSString *deviceTokenString = [AWSPinpointEndpointProfile hexStringFromData:tokenData];
-        
+        if ([userDefaults objectForKey:AWSDeviceTokenKey]) {
```
[the remark this answers is not in the document]
Comment (phantumcode) [reply AWSPinpoint/AWSPinpointEndpointProfile.m]: Good suggestions! I've updated AWSPinpointNotificationManager's `interceptDidRegisterForRemoteNotificationsWithDeviceToken` and made the suggested rewrite in `AWSPinpointEndpointProfile`

File: AWSiOSSDKv2.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -1648,6 +1648,13 @@
 			remoteGlobalIDString = CE0D416C1C6A66E5006B91B5;
 			remoteInfo = AWSCore;
 		};
+		891417A82880C00900C61658 /* PBXContainerItemProxy */ = {
```
[the remark this answers is not in the document]
Comment (phantumcode) [reply AWSiOSSDKv2.xcodeproj/project.pbxproj:1651]: Yes, the integration test needs a test app host in order to access the keychain for read/writes

File: AWSPinpoint/AWSPinpointTargetingClient.m [hunk 3]
```diff
@@ -143,6 +174,43 @@ - (void) addMetricsAndAttributesToEndpointProfile:(AWSPinpointEndpointProfile *)
...
+    }
+    
+    if ([userDefaults objectForKey:AWSPinpointEndpointAttributesKey] != nil) {
+        NSMutableDictionary *attributes = [userDefaults objectForKey:AWSPinpointEndpointAttributesKey];
+        NSError *codingError;
+        NSData *attributesData = [AWSNSCodingUtilities versionSafeArchivedDataWithRootObject:attributes
+                                                                            requiringSecureCoding:YES
+                                                                                            error:&codingError];
+        if (codingError) {
+            AWSDDLogError(@"Error archiving global attributes data. Updating service but not persisting locally: %@", codingError);
```
[the remark this answers is not in the document]
Comment (phantumcode) [reply AWSPinpoint/AWSPinpointTargetingClient.m]: Updated log message and logic to not persist the wrong values

File: AWSPinpoint/AWSPinpointTargetingClient.m [hunk 4]
```diff
@@ -207,8 +275,14 @@ - (void)addAttribute:(NSArray *)theValue
...
         //Save value to disk
         [self.globalAttributes setValue:theValue forKey:theKey];
-        [self.context.configuration.userDefaults setObject:self.globalAttributes forKey:AWSPinpointEndpointAttributesKey];
-        [self.context.configuration.userDefaults synchronize];
+        NSError *codingError;
+        NSData *globalAttributesData = [AWSNSCodingUtilities versionSafeArchivedDataWithRootObject:self.globalAttributes
+                                                                            requiringSecureCoding:YES
+                                                                                            error:&codingError];
+        if (codingError) {
+            AWSDDLogError(@"Error archiving globalAttributesData. Updating service but not persisting locally: %@", codingError);
```
Comment (ruisebas) [inline AWSPinpoint/AWSPinpointTargetingClient.m:283]: This is still persisting even though there's an error

File: AWSPinpoint/AWSPinpointTargetingClient.m [hunk 5]
```diff
@@ -221,8 +295,14 @@ - (void)removeAttributeForKey:(NSString *) theKey {
...
     @synchronized(self) {
         [self.globalAttributes removeObjectForKey:theKey];
-        [self.context.configuration.userDefaults setObject:self.globalAttributes forKey:AWSPinpointEndpointAttributesKey];
-        [self.context.configuration.userDefaults synchronize];
+        NSError *codingError;
+        NSData *globalAttributesData = [AWSNSCodingUtilities versionSafeArchivedDataWithRootObject:self.globalAttributes
+                                                                            requiringSecureCoding:YES
+                                                                                            error:&codingError];
+        if (codingError) {
+            AWSDDLogError(@"Error archiving globalAttributesData. Updating service but not persisting locally: %@", codingError);
```
Comment (ruisebas) [inline AWSPinpoint/AWSPinpointTargetingClient.m:303]: This is still persisting even though there's an error

File: AWSPinpoint/AWSPinpointTargetingClient.m [hunk 6]
```diff
@@ -243,8 +323,14 @@ - (void)addMetric:(NSNumber *)theValue
...
     @synchronized(self) {
         [self.globalMetrics setValue:theValue forKey:theKey];
-        [self.context.configuration.userDefaults setObject:self.globalMetrics forKey:AWSPinpointEndpointMetricsKey];
-        [self.context.configuration.userDefaults synchronize];
+        NSError *codingError;
+        NSData *globalMetricsData = [AWSNSCodingUtilities versionSafeArchivedDataWithRootObject:self.globalMetrics
+                                                                            requiringSecureCoding:YES
+                                                                                            error:&codingError];
+        if (codingError) {
+            AWSDDLogError(@"Error archiving globalMetricsData. Updating service but not persisting locally: %@", codingError);
```
Comment (ruisebas) [inline AWSPinpoint/AWSPinpointTargetingClient.m:331]: This is still persisting even though there's an error

File: AWSPinpoint/AWSPinpointTargetingClient.m [hunk 7]
```diff
@@ -257,8 +343,14 @@ - (void)removeMetricForKey:(NSString *) theKey {
...
     @synchronized(self) {
         [self.globalMetrics removeObjectForKey:theKey];
-        [self.context.configuration.userDefaults setObject:self.globalMetrics forKey:AWSPinpointEndpointMetricsKey];
-        [self.context.configuration.userDefaults synchronize];
+        NSError *codingError;
+        NSData *globalMetricsData = [AWSNSCodingUtilities versionSafeArchivedDataWithRootObject:self.globalMetrics
+                                                                            requiringSecureCoding:YES
+                                                                                            error:&codingError];
+        if (codingError) {
+            AWSDDLogError(@"Error archiving globalMetricsData. Updating service but not persisting locally: %@", codingError);
```
Comment (ruisebas) [inline AWSPinpoint/AWSPinpointTargetingClient.m:351]: This is still persisting even though there's an error

**Review by ruisebas — CHANGES_REQUESTED** [review]

File: AWSPinpoint/AWSPinpointEndpointProfile.m [hunk 8]
```diff
@@ -64,22 +65,32 @@ - (instancetype) initWithApplicationId:(NSString*) applicationId
...
+            [keychain setData:tokenData forKey:AWSDeviceTokenKey];
+            [userDefaults removeObjectForKey:AWSDeviceTokenKey];
+            [userDefaults synchronize];
+        } else {
+            tokenData = [keychain dataForKey:AWSDeviceTokenKey];
+        }
+
         NSString *deviceTokenString = [AWSPinpointEndpointProfile hexStringFromData:tokenData];
-        
+        _address = deviceTokenString;
```
[the remark this answers is not in the document]
Comment (phantumcode) [reply AWSPinpoint/AWSPinpointEndpointProfile.m]: Updated to remove unnecessary line change :)

File: AWSPinpoint/AWSPinpointNotificationManager.m [hunk 9]
```diff
@@ -113,10 +113,9 @@ - (BOOL)interceptDidFinishLaunchingWithOptions:(nullable NSDictionary *)launchOp
 
 - (void)interceptDidRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken {
     //Check if device token has changed
-    NSData *currentToken = [self.context.configuration.userDefaults objectForKey:AWSDeviceTokenKey];
+    NSData *currentToken = [self.context.keychain dataForKey:AWSDeviceTokenKey];
     if (![currentToken isEqualToData:deviceToken]) {
-        [self.context.configuration.userDefaults setObject:deviceToken forKey:AWSDeviceTokenKey];
```
[the remark this answers is not in the document]
Comment (phantumcode) [reply AWSPinpoint/AWSPinpointNotificationManager.m:118]: The migration from user defaults to keychain occurs at init of `AWSPinpointEndpointProfile`, so by the time `interceptDidRegisterForRemoteNotificationsWithDeviceToken` is called, the migration should already be executed and completed.  As a safeguard, I've updated the migration logic to check if a newer device token already exists in the keychain, if it does then throw away the outdated device token in the keychain.

File: AWSPinpoint/AWSPinpointTargetingClient.m (hunk 4 above)
Comment (phantumcode) [reply AWSPinpoint/AWSPinpointTargetingClient.m]: Missed it, should be fixed now.

File: AWSPinpoint/AWSPinpointTargetingClient.m (hunk 5 above)
Comment (phantumcode) [reply AWSPinpoint/AWSPinpointTargetingClient.m]: Missed it, should be fixed now.

File: AWSPinpoint/AWSPinpointTargetingClient.m (hunk 6 above)
Comment (phantumcode) [reply AWSPinpoint/AWSPinpointTargetingClient.m]: Missed it, should be fixed now.

File: AWSPinpoint/AWSPinpointTargetingClient.m (hunk 7 above)
Comment (phantumcode) [reply AWSPinpoint/AWSPinpointTargetingClient.m]: Missed it, should be fixed now.

File: AWSPinpoint/AWSPinpointTargetingClient.m [hunk 10]
```diff
@@ -74,9 +74,9 @@ - (instancetype) initWithContext:(AWSPinpointContext *) context {
 
 - (void)initGlobalAttributes {
     NSData *customAttributesData = [_keychain dataForKey:AWSPinpointEndpointAttributesKey];
-    if (customAttributesData == nil) {
-        _globalAttributes = [[NSMutableDictionary alloc] init];
-    } else {
+    NSMutableDictionary *attributes = [_context.configuration.userDefaults objectForKey:AWSPinpointEndpointAttributesKey];
```
Comment (ruisebas) [inline AWSPinpoint/AWSPinpointTargetingClient.m:77]: nit: I'd suggest use a more descriptive name that reflects what's going on, like `legacyStoredAttributes` or similar

File: AWSPinpoint/AWSPinpointTargetingClient.m [hunk 11]
```diff
@@ -85,31 +85,47 @@ - (void)initGlobalAttributes {
...
+        _globalAttributes = [[NSMutableDictionary alloc] init];
     }
 }
 
 - (void) initGlobalMetrics {
     NSData *customMetricsData = [_keychain dataForKey:AWSPinpointEndpointMetricsKey];
-    if (customMetricsData == nil) {
-        _globalMetrics = [[NSMutableDictionary alloc] init];
-    } else {
+    NSMutableDictionary *metrics = [_context.configuration.userDefaults objectForKey:AWSPinpointEndpointMetricsKey];
```
Comment (ruisebas) [inline AWSPinpoint/AWSPinpointTargetingClient.m:102]: Same as before, consider renaming

**Review by ruisebas — APPROVED** [review]
Left some nit comments, but you can just merge whether you address them or not :)

**Review by ruisebas — APPROVED** [review]

**Pull request merged**